p2p/dhtdiscovery/discovery_test.go

```diff
@@ -62,7 +57,7 @@ func TestSanity(t *testing.T) {
 	)
 	require.NoError(t, err)
 	bootdisc.Start()
-	defer bootdisc.Stop()
+	t.Cleanup(bootdisc.Stop)
```

Comment:
In my opinion `defer` is correct here. Its primarily a conceptual difference `defers` are executed just before the test ends (before `Test_XXX` returns), while `t.Cleanup` are executed after the test completes.

The question here is when should `bootdisc` be stopped? As part of the test or as part of the cleanup?

Reply:
Well on one hand `bootdisc.Stop()` belongs to the test
But re `t.Cleanup()`, the test will still fail if it fails, but it makes refactoring the test later much easier in most cases.
If you extract the piece of code into a "setup" function, it is very easy to shoot yourself in the foot with `defer`s, b/c they'll stop the services etc. immediately before the setup function returns. With `t.Cleanup`, this doesn't happen.
Also, `t.Cleanup()` does the right thing if subtests are added to the test later and these subtests use `t.Parallel()`

Reply:
(the latter thing about `t.Parallel()` is probably not exactly applicable to this particular case but still worth noting)

Reply:
In a setup method I'd expect `bootdisc.Stop` to be called in a cleanup method that is returned by the setup method and deferred.

It depends on the specific case, but for parallel tests it's also imo safer to use `defer`:

```go
func TestMultiple(t *testing.T) {
	t.Run("test1", func (t *testing.T) {
		t.Parallel()

		data, cleanup := someSetup(t)
		defer cleanup()
		// Actual test
	})
	t.Run("test2", func (t *testing.T) {
		t.Parallel()

		data, cleanup := someSetup(t)
		defer cleanup()
		// Actual test
	})
}
```

if those `cleanup` calls are invoked with `t.Cleanup` it could happen by accident that they are using the `testing.T` of the main test which is less likely with `defer`.

Another reason to use `defer` over `t.Cleanup` is to avoid issues on Windows: if `t.Cleanup(bootdisc.Stop())` closes file handles it works on macos and linux but on windows it will fail with an error that go cannot delete the temporary files because they have open file handles.

Reply:
Ok, I can change it, although it was more than once that I got bitten by `defer` during test refactoring and spent some time debugging the problem afterwards, as it is not always very obvious until you spot that `defer`

In parallel tests, the setup outside the subtests might use `t.Cleanup()` while setting up common slow-to-start fixture for the subtests. In this case the cleanup code must use "main" `testing.T`, and the cleanups will only get executed after all of the subtests finish:

```
func TestMultiple(t *testing.T) {
	data := someSetup(t) // uses t.Cleanup() for cleanup code
	t.Run("test1", func (t *testing.T) {
		t.Parallel()
		// Actual test
	})
	t.Run("test2", func (t *testing.T) {
		t.Parallel()
		// Actual test
	})
	// any cleanup done in `setup` gets executed only after
	// both `test1` and `test2` finish running
}
```

Re the Windows issue, `t.TempDir()` uses `t.Cleanup()` internally to delete the directory. Cleanups are executed strictly in reverse order, just like `defer`, so the only possibility I see is scheduling the cleanup func that closes the file _before_ invoking `t.TempDir()` somehow, which is a mistake by itself

Reply:
More re parallel: https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests

Reply:
> More re parallel: https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests

To me this is kinda expected? `defer` is called before the test ends (so sub tests might still be running) while `t.Cleanup` is called after.

> Re the Windows issue, `t.TempDir()` uses `t.Cleanup()` internally to delete the directory. Cleanups are executed strictly in reverse order, just like `defer`, so the only possibility I see is scheduling the cleanup func that closes the file before invoking `t.TempDir()` somehow, which is a mistake by itself

That was not the issue. The issue was that the file wasn't closed when the cleanup happened, and to make sure that the close happened before the delete was to `defer` the call rather than to pass it to `t.Cleanup`. I do remember some tests where we had to change `t.Cleanup` to `defer` or it wouldn't pass on windows because of that very reason. I wasn't able to quickly come up with a simple example. Maybe this issue is also not present any more in Go 1.22

```